_posts/2019/2019-01-13-use-multiple-migrations-when-adding-database-constraints.md

```diff
@@ -0,0 +1,95 @@
+---
+title: Use Multiple Migrations When Adding Database Constraints
+description: You need to avoid a long table rebuild and a write-lock
+layout: article
+category: ruby
+image:
+  base: '2018/use-multiple-migrations-when-adding-database-constraints'
+  alt: 'Locks'
+  source: 'https://unsplash.com/photos/8_NI1WTqCGY'
+  credit: 'Marcos Mayer'
+---
+
+Adding constraints to your application at the database layer is a good idea as it provides an extra layer of quality control on top of the data powering your application.
+
+One way to do that is to add default values or constraints (like making sure a field can’t be blank), this is easy to do with Rails’ migrations. But beware...
+
+
+## Instead of…
+
+…adding a column and stipulating a default non-null value all in one go.
+
+```ruby
+class AddComposerToSongs < ActiveRecord::Migration[5.2]
+  def change
+    add_column :songs, :composer, :string, default: "Lin-Manuel Miranda", null: false
+  end
+end
+```
+
+
+## Use…
+
+…a multiple migration strategy to add a field with a constraint to your databases
+
+
+### Add a column
+
+```ruby
+class AddComposerToSongs < ActiveRecord::Migration[5.2]
+  def change
+    add_column :songs, :composer, :string
+  end
+end
+```
+
+
+### Deploy code that sets default, in Ruby
+
+```ruby
+class Song < ApplicationRecord
+  before_save :set_default_composer
+
+  def set_default_composer
+    self.composer = "Lin-Manuel Miranda" if composer.nil?
+  end
+end
+```
+
+
+### Update records with nil values
+
+```ruby
+Song.where(composer: nil).update_all(composer: "Lin-Manuel Miranda")
```

Comment:
Is this on the command line?

Reply:
Also, if you're doing this post the migration, why do you need the `Deploy code that sets default, in Ruby` section.

Why can't you do:
1. The first migration
1. Then update the records with `nil` values
1. Then change the column to have a constraint?

Am I missing something? Is it in case there are writes to the DB in between steps 1 and 2?